Global static mutex random_sync causes order of destruction bug

# Issue description

When using zmqpp with libzmq 4.2.3 there is an order of destruction bug in static builds if there are contexts defined globally/static. This is due to the global static mutex random_sync in random.cpp being destructed before the context is destructed.

I think the better approach would be to have client application control the initialization of sodium_init() via the zmq::random_open()/close() function. Client code would need to call this at start/end of main(). I think adding a couple functions like zmq_init() and zmq_destroy() to wrap these kind of init calls would do the trick.

# Environment

* libzmq version (commit hash if unreleased): 4.2.3
* OS: Ubuntu 16.04

# What's the actual result? (include assertion message & call stack if applicable)

```
Program received signal SIGABRT, Aborted.
0x00007ffff5534428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) list
49	in ../sysdeps/unix/sysv/linux/raise.c
(gdb) bt
#0  0x00007ffff5534428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff553602a in __GI_abort () at abort.c:89
#2  0x00000000012f39ee in zmq::zmq_abort (errmsg_=errmsg_@entry=0x7ffff568c10a "Invalid argument") at src/err.cpp:87
#3  0x00000000012fde65 in zmq::mutex_t::lock (this=0x4431200 <random_sync>) at src/mutex.hpp:122
#4  zmq::scoped_lock_t::scoped_lock_t (mutex_=..., this=<synthetic pointer>) at src/mutex.hpp:168
#5  zmq::random_close () at src/random.cpp:110
#6  0x00000000012ee546 in zmq::ctx_t::~ctx_t (this=0x61600000f980, __in_chrg=<optimized out>) at src/ctx.cpp:124
#7  0x00000000012eeb65 in zmq::ctx_t::terminate (this=0x61600000f980) at src/ctx.cpp:201
#8  0x00000000012dc711 in zmqpp::context::terminate() ()
#9  0x0000000000485eef in zmqpp::context::~context (this=0x4431188 <zmqpp::actor::actor_pipe_ctx_>, __in_chrg=<optimized out>) at ../../../../trunk/thirdparty/zmqpp4.2.0/include/zmqpp/context.hpp:100
#10 0x00007ffff5538ff8 in __run_exit_handlers (status=0, listp=0x7ffff58c35f8 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true) at exit.c:82
#11 0x00007ffff5539045 in __GI_exit (status=<optimized out>) at exit.c:104
#12 0x00007ffff551f837 in __libc_start_main (main=0x4792c0 <main(int, char**)>, argc=2, argv=0x7fffffffdb58, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffdb48) at ../csu/libc-start.c:325
#13 0x00000000004463a9 in _start ()
```


# What's the expected result?

Correct order of destruction and no abort being thrown.

The problem is actually in libzmq because if any client code uses it, and defines a context globally will run into this behaviour, because the order of destruction is undefined and that mutex will always have a possibility of being destroyed before the context.
The zmqpp library can only get around it by somehow removing the need to have a context defined globally. I have reported this issue on their repo and will see if a workaround can be found for the time being.

By the way, this wasn't an issue in libzmq 4.2.2.

________________________________
From: Luca Boccassi <notifications@github.com>
Sent: Sunday, February 4, 2018 3:15:10 PM
To: zeromq/libzmq
Cc: Aaron Zinghini; Author
Subject: Re: [zeromq/libzmq] Global static mutex random_sync causes order of destruction bug (#2915)


I think the better approach would be to have client application control the initialization of sodium_init() via the zmq::random_open()/close() function. Client code would need to call this at start/end of main(). I think adding a couple functions like zmq_init() and zmq_destroy() to wrap these kind of init calls would do the trick.

Sorry, but we cannot do this - it would be an API breakage.

Given the problem is in zmqpp, is there any way to solve it there? Can we fix the ordering?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/zeromq/libzmq/issues/2915#issuecomment-362944671>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AT0vbiu6Nw1OEMwBs_Ym41vAg8uRL6Niks5tRivugaJpZM4R3mwH>.


Aaron Zinghini
Guardian Software Architect










T       +1 917 346 2993
M       +1 917 346 2993
E       aaron.zinghini@seeingmachines.com
W       www.seeingmachines.com<http://www.seeingmachines.com>

[Seeing Machines]<https://www.seeingmachines.com/>

This email is confidential. If you are not the intended recipient, destroy all copies and do not disclose or use the information within. No warranties are given that this email does not contain viruses or harmful code.

--- a/src/random.cpp
+++ b/src/random.cpp
@@ -79,41 +79,76 @@ uint32_t zmq::generate_random ()
 //  The context class cannot be used with static variables as the curve
 //  utility APIs like zmq_curve_keypair also call into the crypto
 //  library.
-//  The safest solution for all use cases therefore is to have a global,
+//  The safest solution for all use cases therefore is to have a
 //  static lock to serialize calls into an initialiser and a finaliser,
 //  using refcounts to make sure that a thread does not close the library
-//  while another is still using it.
+//  while another is still using it. To avoid the static initialization
+//  order fiasco, this is done using function-local statics, if the
+//  compiler implementation supports thread-safe initialization of those.
+//  Otherwise, we fall back to global statics.
+
+//  TODO if there is some other user of libsodium besides libzmq, this must
+//    be synchronized by the application. This should probably also be
+//    configurable via config.h
+
+//  TODO this should probably be done via config.h
+#if __cplusplus >= 201103L                                                     \
+  || (defined(__cpp_threadsafe_static_init)                                    \
+      && __cpp_threadsafe_static_init >= 200806)                               \
+  || (defined(_MSC_VER) && _MSC_VER >= 1900)
+#define ZMQ_HAVE_THREADSAFE_STATIC_LOCAL_INIT 1
+//  TODO this might probably also be set if a sufficiently recent gcc is used
+//  without -fno-threadsafe-statics, but this cannot be determined at
+//  compile-time, so it must be set via config.h
+#else
+#define ZMQ_HAVE_THREADSAFE_STATIC_LOCAL_INIT 0
+#endif
+
+#if !ZMQ_HAVE_THREADSAFE_STATIC_LOCAL_INIT                                     \
+  && (defined(ZMQ_USE_LIBSODIUM)                                               \
+      || (defined(ZMQ_USE_TWEETNACL) && !defined(ZMQ_HAVE_WINDOWS)             \
+          && !defined(ZMQ_HAVE_GETRANDOM)))
 static unsigned int random_refcount = 0;
 static zmq::mutex_t random_sync;
+#endif
 
-void zmq::random_open (void)
+static void manage_random (bool init)
 {
 #if defined(ZMQ_USE_LIBSODIUM)                                                 \
   || (defined(ZMQ_USE_TWEETNACL) && !defined(ZMQ_HAVE_WINDOWS)                 \
       && !defined(ZMQ_HAVE_GETRANDOM))
-    scoped_lock_t locker (random_sync);
 
-    if (random_refcount == 0) {
-        int rc = sodium_init ();
-        zmq_assert (rc != -1);
-    }
-
-    ++random_refcount;
-#else
-    LIBZMQ_UNUSED (random_refcount);
+#if ZMQ_HAVE_THREADSAFE_STATIC_LOCAL_INIT
+    static int random_refcount = 0;
+    static zmq::mutex_t random_sync;
 #endif
+
+    if (init) {
+        zmq::scoped_lock_t locker (random_sync);
+
+        if (random_refcount == 0) {
+            int rc = sodium_init ();
+            zmq_assert (rc != -1);
+        }
+
+        ++random_refcount;
+    } else {
+        zmq::scoped_lock_t locker (random_sync);
+        --random_refcount;
+
+        if (random_refcount == 0) {
+            randombytes_close ();
+        }
+    }
+#endif
+}
+
+void zmq::random_open (void)
+{
+    manage_random (true);
 }
 
 void zmq::random_close (void)
 {
-#if defined(ZMQ_USE_LIBSODIUM)                                                 \
-  || (defined(ZMQ_USE_TWEETNACL) && !defined(ZMQ_HAVE_WINDOWS)                 \
-      && !defined(ZMQ_HAVE_GETRANDOM))
-    scoped_lock_t locker (random_sync);
-    --random_refcount;
-
-    if (random_refcount == 0) {
-        randombytes_close ();
-    }
-#endif
+    manage_random (false);
 }

